File: python/pytorch/models/resnet18.py
import os
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import models, transforms
from torch.utils.data import DataLoader
from torchvision.datasets import ImageFolder
from tqdm import tqdm
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[3]))
from tools.deepgreen_bench import Harness, RunContext


def build_resnet18(num_classes: int = 100, pretrained: bool = False) -> nn.Module:
    model = models.resnet18(pretrained=pretrained)
    model.fc = nn.Linear(model.fc.in_features, num_classes)

    """if not pretrained:
        def init_weights(m):
            if isinstance(m, nn.Conv2d) or isinstance(m, nn.Linear):
                nn.init.kaiming_normal_(m.weight)
                if m.bias is not None:
                    nn.init.zeros_(m.bias)

        model.apply(init_weights)"""

    return model


def get_loaders(dataset_path, batch_size=128, img_size=(32, 32), grayscale=False, test_split="test"):
    transform_list = [transforms.Resize(img_size)]
    if grayscale:
        transform_list.append(transforms.Grayscale(num_output_channels=3))
    transform_list.append(transforms.ToTensor())

    transform = transforms.Compose(transform_list)
    # Tiny ImageNet ships train/ and val/, the other two ship train/ and test/.
    # TensorFlow and JAX already resolved this at runtime; this stack took the
    # split from a caller-supplied default and failed on Tiny ImageNet.
    if not os.path.isdir(os.path.join(dataset_path, test_split)):
        for candidate in ("test", "val"):
            if os.path.isdir(os.path.join(dataset_path, candidate)):
                test_split = candidate
                break
        else:
            raise FileNotFoundError(
                f"no test or val split under {dataset_path}")
    train_set = ImageFolder(root=os.path.join(dataset_path, "train"), transform=transform)
    test_set = ImageFolder(root=os.path.join(dataset_path, test_split), transform=transform)

    train_loader = DataLoader(train_set, batch_size=batch_size, shuffle=True, num_workers=2)
    test_loader = DataLoader(test_set, batch_size=batch_size, shuffle=False, num_workers=2)

    return train_loader, test_loader, len(train_set.classes)


def train(model, train_loader, criterion, optimizer, device):
    model.train()
    running_loss = 0.0
    for inputs, targets in train_loader:
        inputs, targets = inputs.to(device), targets.to(device)
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, targets)
        loss.backward()
        optimizer.step()
        running_loss += loss.item() * inputs.size(0)
    return running_loss / len(train_loader.dataset)


def evaluate(model, test_loader, criterion, device):
    model.eval()
    total, correct = 0, 0
    loss_sum = 0.0
    with torch.no_grad():
        for inputs, targets in test_loader:
            inputs, targets = inputs.to(device), targets.to(device)
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            loss_sum += loss.item() * inputs.size(0)
            _, predicted = outputs.max(1)
            total += targets.size(0)
            correct += predicted.eq(targets).sum().item()
    acc = 100. * correct / total
    return loss_sum / total, acc


def run_experiment(dataset_path, output_file, checkpoint_path, img_size=(32, 32), grayscale=False,
                   test_split="test", epochs=30, batch_size=128, repetition=0, seed=None,
                   dataset_name=None, precision="fp32"):
    """Run one *independent* repetition of the resnet18 experiment.

    Changes with respect to the first campaign:
      * CodeCarbon is configured through tools.deepgreen_bench so that every
        ecosystem shares one tracking mode, sampling interval and version
        requirement (reviewer 1 comments 8 and 9).
      * ``repetition`` and ``seed`` make run-level replication possible; the
        first campaign executed each configuration once and averaged epochs,
        which are not independent (reviewer 1 comment 5, reviewer 3 major
        comment 2).
      * Test accuracy and both losses are persisted per epoch, so energy can be
        normalised by the useful work produced (reviewer 1 comment 6,
        reviewer 3 major comment 3).
      * Precision is pinned rather than left to the framework default
        (reviewer 1 comment 8).
    """
    ctx = RunContext(
        ecosystem="Python/PyTorch",
        model="resnet18",
        dataset=dataset_name or Path(dataset_path).name,
        repetition=repetition,
        seed=seed,
        epochs=epochs,
        batch_size=batch_size,
        precision=precision,
    )

    with Harness(ctx) as bench:
        bench.set_seeds()

        device = torch.device(
            "cuda" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "cpu")
        train_loader, test_loader, num_classes = get_loaders(
            dataset_path, batch_size, img_size, grayscale, test_split)

        model = build_resnet18(num_classes=num_classes, pretrained=False)
        model.to(device)

        criterion = nn.CrossEntropyLoss()
        optimizer = optim.Adam(model.parameters(), lr=1e-4)

        for epoch in range(1, epochs + 1):
            print(f"Epoch {epoch}/{epochs} (repetition {ctx.repetition}, seed {ctx.seed})")

            with bench.track("train", epoch):
                train_loss = train(model, train_loader, criterion, optimizer, device)

            with bench.track("eval", epoch):
                test_loss, test_acc = evaluate(model, test_loader, criterion, device)

            bench.log_metrics(
                epoch,
                train_loss=train_loss,
                test_loss=test_loss,
                test_acc=test_acc,
            )
            print(f"Train Loss={train_loss:.4f}, Test Loss={test_loss:.4f}, Test Acc={test_acc:.2f}%")

        os.makedirs(os.path.dirname(checkpoint_path) or "checkpoints", exist_ok=True)
        torch.save(model.state_dict(), checkpoint_path)

File: R/models/resnet18.r
# R/models/resnet18.r
library(torch)
library(torchvision)
library(coro)

# --- tracking utility (CodeCarbon via Python CLI) ---
# Shared measurement bridge (spec S5). The stack-private helper set
# tracking_mode = "process", the only one in the campaign to do so, which
# excluded almost all host energy and made its totals incomparable.
source("R/scripts/deepgreen_tracking.r")

# ===== Helpers (log) =====
.now <- function() format(Sys.time(), "%Y-%m-%d %H:%M:%S")
.log <- function(...) cat(sprintf("[%s] ", .now()), sprintf(...), "\n")

# ===== Modello =====
build_resnet18 <- function(num_classes = 100, pretrained = FALSE) {
  model <- torchvision::model_resnet18(pretrained = pretrained)
  in_f <- model$fc$in_features
  model$fc <- nn_linear(in_f, num_classes)
  model
}

# ===== Data loaders =====
get_loaders <- function(dataset_path, batch_size = 128, img_size = c(32, 32),
                        grayscale = FALSE, test_split = "test") {
  .log("Checking dataset path: %s", dataset_path)
  if (!dir.exists(file.path(dataset_path, "train"))) {
    stop("Train directory does not exist: ", file.path(dataset_path, "train"))
  }
  if (!dir.exists(file.path(dataset_path, test_split))) {
    stop("Test directory does not exist: ", file.path(dataset_path, test_split))
  }

  # Fashion MNIST is grayscale (1 channel), but ResNet18 expects 3 channels
  transform <- function(img) {
    # transform_to_tensor, as R/VGG-16 and every other stack does it. This read
    # `torch_tensor(img / 255)`, and image_folder_dataset's default loader
    # already returns values in [0, 1] -- measured: base_loader on a campaign
    # PNG gives an array with range 0 to 1. So R/ResNet-18 trained on [0, 1/255]
    # while R/VGG-16 trained on [0, 1]: one ecosystem, two input pipelines, and
    # S3 fixes the pipeline for all stacks with no check covering it. Largely
    # absorbed by the BatchNorm after conv1, which is why it never showed in the
    # accuracies -- and why nothing but reading the source would have found it.
    img <- torchvision::transform_to_tensor(img)

    # Resize to target size
    img <- torchvision::transform_resize(img, size = c(img_size[2], img_size[1]))

    # If grayscale, replicate the single channel to 3 channels
    if (grayscale) {
      if (img$size(1) == 1) {
        img <- img$repeat_interleave(3, dim = 1) # Replicate to [3, height, width]
      }
    }
    
    img
  }

  .log("Loading train dataset from %s", file.path(dataset_path, "train"))
  train_set <- tryCatch({
    torchvision::image_folder_dataset(file.path(dataset_path, "train"), transform = transform)
  }, error = function(e) {
    stop("Failed to load train dataset: ", e$message)
  })
  .log("Train dataset loaded. Classes: %s, Samples: %d", paste(train_set$classes, collapse = ", "), length(train_set))

  .log("Loading test dataset from %s", file.path(dataset_path, test_split))
  test_set <- tryCatch({
    torchvision::image_folder_dataset(file.path(dataset_path, test_split), transform = transform)
  }, error = function(e) {
    stop("Failed to load test dataset: ", e$message)
  })
  .log("Test dataset loaded. Classes: %s, Samples: %d", paste(test_set$classes, collapse = ", "), length(test_set))

  train_loader <- dataloader(train_set, batch_size = batch_size, shuffle = TRUE, num_workers = 2)  # was 0: single-threaded decoding made R ~11x slower than Rust
  test_loader  <- dataloader(test_set, batch_size = batch_size, shuffle = FALSE, num_workers = 2)  # was 0: single-threaded decoding made R ~11x slower than Rust

  list(
    train_loader = train_loader,
    test_loader = test_loader,
    num_classes = length(train_set$classes)
  )
}
# ===== Train / Eval =====
train <- function(model, train_loader, criterion, optimizer, device) {
  model$train()
  running_loss <- 0
  coro::loop(for (b in train_loader) {
    inputs <- b[[1]]$to(device = device)
    targets <- b[[2]]$to(device = device)
    optimizer$zero_grad()
    outputs <- model(inputs)
    loss <- criterion(outputs, targets)
    loss$backward()
    optimizer$step()
    running_loss <- running_loss + loss$item() * inputs$size(1)
  })
  running_loss / length(train_loader$dataset)
}

evaluate <- function(model, test_loader, criterion, device) {
  model$eval()
  total <- 0; correct <- 0; loss_sum <- 0
  with_no_grad({
    coro::loop(for (b in test_loader) {
      inputs <- b[[1]]$to(device = device)
      targets <- b[[2]]$to(device = device)
      outputs <- model(inputs)
      loss <- criterion(outputs, targets)
      loss_sum <- loss_sum + loss$item() * inputs$size(1)
      pred <- torch_max(outputs, dim = 2)[[2]]
      total <- total + targets$size(1)
      correct <- correct + (pred == targets)$sum()$item()
    })
  })
  list(loss = loss_sum / total, acc = 100 * correct / total)
}


# ===== Shared TorchScript module (spec S1) =====
# The four LibTorch-based ecosystems must train the SAME module. In the first
# campaign this stack built torchvision-for-R's own ResNet-18/VGG-16 while C++
# and Rust each used a hand-written port and Python used torchvision, so the
# "one backend, four bindings" control group actually compared four
# implementations. The module is produced once by
# scripts/export_torchscript_models.py; the exporting torch build must match the
# LibTorch that the R torch package links against.
deepgreen_model_path <- function(arch, dataset) {
  root <- Sys.getenv("DEEPGREEN_MODELS", unset = "models")
  file.path(root, paste0(arch, "_", dataset, ".pt"))
}

deepgreen_dataset_key <- function(dataset_path) {
  base <- basename(dataset_path)
  switch(sub("_png$", "", base),
         "cifar100"      = "cifar100",
         "fashion_mnist" = "fashionmnist",
         "tiny_imagenet" = "tinyimagenet200",
         stop(sprintf("unknown dataset directory: %s", base)))
}

load_shared_module <- function(arch, dataset_path, device) {
  path <- deepgreen_model_path(arch, deepgreen_dataset_key(dataset_path))
  if (!file.exists(path)) {
    stop(sprintf(
      "shared TorchScript module not found at %s; run scripts/export_torchscript_models.py",
      path))
  }
  m <- torch::jit_load(path)
  m$to(device = device)
  m
}

# ===== Esperimento (con tracking integrato) =====
run_experiment <- function(dataset_path, checkpoint_path,
                           img_size = c(32, 32), grayscale = FALSE, test_split = "test",
                           epochs = 30, batch_size = 128,
                           run_id = NULL, python_bin = Sys.getenv("PYTHON_BIN", unset = "python")) {

  params <- dg_run_params()
  epochs <- params$epochs
  torch_manual_seed(params$seed)
  set.seed(params$seed)

  device <- if (cuda_is_available()) torch_device("cuda") else torch_device("cpu")
  if (device$type != "cuda") {
    stop("R/torch does not see a GPU. Measuring now would attribute a CPU fallback ",
         "to the ecosystem; set DEEPGREEN_ALLOW_CPU=1 to override deliberately.")
  }
  .log("device=%s | repetition=%s seed=%s epochs=%s",
       device$type, params$repetition, params$seed, epochs)

  loaders <- get_loaders(dataset_path, batch_size, img_size, grayscale, test_split)
  # R/torch is the one LibTorch binding that cannot use the shared TorchScript
  # module (spec S1). In torch 0.17.0 a script_module's $train() and $eval()
  # raise "unused argument", and the underlying handle is not reachable through
  # the documented fields, so the module cannot be switched between training and
  # evaluation. The shared modules are exported in training mode, so this stack
  # would evaluate with batch norm using batch statistics -- the same defect
  # found in the TensorFlow and Rust stacks. It therefore builds its own model,
  # and the architecture parity that holds for Python, C++ and Rust does not
  # hold here. See results/analysis/experiment_spec.md, S1.
  model <- build_resnet18(num_classes = loaders$num_classes, pretrained = FALSE)
  dg_assert_params(model)
  model$to(device = device)
  criterion <- nn_cross_entropy_loss()
  optimizer <- optim_adam(model$parameters, lr = 1e-4)

  # --- init energy tracker (CLI) ---
  dataset_name <- basename(dataset_path)  # es: "cifar100_png"
  dg_init()
  dg_datafp(loaders$test_loader)

  on.exit(dg_shutdown(), add = TRUE)

  # crea la cartella del checkpoint finale se manca
  dir.create(dirname(checkpoint_path), showWarnings = FALSE, recursive = TRUE)

  for (epoch in 1:epochs) {
    .log("Epoch %d/%d", epoch, epochs)

    # ---- TRAIN (tracciato) ----
    dg_start("train", epoch)
    train_loss <- train(model, loaders$train_loader, criterion, optimizer, device)
    dg_stop()

    # ---- EVAL (tracciato) ----
    dg_start("eval", epoch)
    eval <- evaluate(model, loaders$test_loader, criterion, device)
    dg_stop()

    # Outside the tracked window: writing the metric must not be measured.
    dg_metric(epoch, train_loss, eval$loss, eval$acc)
    .log("Train Loss=%.4f | Test Loss=%.4f, Acc=%.2f%%",
         train_loss, eval$loss, eval$acc)
  }

  # Salva solo il modello finale
  torch_save(model$state_dict(), checkpoint_path)
  .log("Modello finale salvato in %s", checkpoint_path)
}
